raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -984,11 +991,16 @@ private CompletableFuture<FetchResponseData> handleFetchRequest(
                 Throwable cause = exception instanceof ExecutionException ?
                     exception.getCause() : exception;
 
-                // If the fetch timed out in purgatory, it means no new data is available,
-                // and we will complete the fetch successfully. Otherwise, if there was
-                // any other error, we need to return it.
                 Errors error = Errors.forException(cause);
-                if (error != Errors.REQUEST_TIMED_OUT) {
+                if (error == Errors.REQUEST_TIMED_OUT) {
+                    // Note that for this case the calling thread is the expiration service thread and not the
+                    // polling thread.
+                    //
+                    // If the fetch request timed out in purgatory, it means no new data is available,
+                    // just return the original fetch response.
+                    return response;
```

Comment:
This is probably the simplest fix for now. An alternative might be to use the callback to put an event on the queue that could be handled in the event loop.

Reply:
Yes. That is what I would like to do in the long-term. In the long-term I plan to have a more generic event executor that would allow us to keep the current concurrency model, remove the expiration service thread, and keep all our deterministic unittest and simulation tests.

It is a bigger change and I didn't want to block this fix on that.